### PR TITLE
Replace sine table with 5th order approximation

### PR DESCRIPTION
**Pull request by ndsl7109256**
The previous implementation used the pre-calculated fixed-point
arithmetic table for sine values, which added over 2 KiB to the code
size.

To solve this issue, I implemented the 5th order polynomial
approximation for sine, which reduced the code size from 2556 bytes
to 494 bytes (by about 80%) while maintaining acceptable error (RMS
error is 8.866 observed with 0-90 degrees, 0-1024 in fixed-point
representation)

Original code size with sine table.
```bash
$ size .libtwin.a/src/trig.c.o
   text	   data	    bss	    dec	    hex	filename
   2556	      0	      0	   2556	    9fc	.libtwin.a/src/trig.c.o
```

Code size after using 5th order approximation
```bash
$ size .libtwin.a/src/trig.c.o
   text	   data	    bss	    dec	    hex	filename
   494	      0	      0	   494	    1ee	.libtwin.a/src/trig.c.o
```

I initially tried a 3rd order approximation, but the error was slightly large, causing the drawn circle to appear not perfectly round. Below is some analysis.

image

RMS Error Comparing with Sine table (valued from 0-65536, observed with 0-90 degrees, 0-1024 in fixed-point
representation)
|              | 3rd approximation | 5th approximation |
| ------------ | ----------------- | ----------------- |
| RMS          | 878.000907        | 8.8665040         |

image

ref:
https://www.nullhardware.com/blog/fixed-point-sine-and-cosine-for-embedded-systems/
https://hackmd.io/@rickywu0421/FinalProject#透過-3rd-order-sine-approximation-模擬-RSSI

### CODE REVIEW

**Review by jserv — CHANGES_REQUESTED** [review]
You can implement a new public function, `twin_sincos(twin_angle_t a, twin_fixed_t *sin, twin_fixed_t *cos)`, to compute both sine and cosine at the same time and store the results in `*sin` and `*cos`, as several applications need the sine and cosine of the same angle `a`. Then, you can replace all occurrences of `twin_sin` and `twin_cos` with the shortcut `twin_sincos`, including the code in the files `src/path.c` and `src/matrix.c`.

Reference: https://man7.org/linux/man-pages/man3/sincos.3.html (GNU extension)

**Comment by jserv** [comment]
glibc implementation of `sincos`:
* [sysdeps/ieee754/flt-32/s_sincosf.h](https://github.com/bminor/glibc/blob/master/sysdeps/ieee754/flt-32/s_sincosf.h)
* [sysdeps/ieee754/flt-32/s_sincosf_data.c](https://github.com/bminor/glibc/blob/master/sysdeps/ieee754/flt-32/s_sincosf_data.c)
* [sysdeps/ieee754/flt-32/s_sincosf.c](https://github.com/bminor/glibc/blob/master/sysdeps/ieee754/flt-32/s_sincosf.c)

**Comment by jouae** [comment]
@ndsl7109256 Hey! You do a great job!

Here is a suggestion make your report getting better.

When mentioning RMS or MSE, be sure to include the number of observations.

**Review by jouae — COMMENTED** [review]
In the PR, you wrote about the method used to compare code sizes, third-order and fifth-order approximation tests, and included references. However, in this commit, you did not mention them. Additionally, for the 80%, please add the actual data and the calculation process based on that actual data.

Remember, "writing a good commit" is an ongoing challenge for every engineer, including myself. Commits are written for others to read; they are not just explanations to convince others to accept a PR but also records of references, ways to solve problems, simple comparative data, and the reasons for adopting a particular solution at that time.

Once again, you did a great job, which includes code size comparisons, differences between third-order and fifth-order approximations, and providing data sources. Every piece of data and record is the result of your hard work and should be documented in the commit for future reference to see these contributions.

Thank you.

File: include/twin.h [hunk 1]
```diff
@@ -736,6 +736,8 @@ void twin_path_rmove(twin_path_t *path, twin_fixed_t x, twin_fixed_t y);
 
 void twin_path_draw(twin_path_t *path, twin_fixed_t x, twin_fixed_t y);
 
+void twin_path_draw_polar(twin_path_t *path, twin_angle_t deg);
```
Comment (jserv) [inline include/twin.h:739]: This function should be internal. i.e., don't declare in the public header.

File: src/path.c [hunk 2]
```diff
@@ -134,6 +134,13 @@ void twin_path_draw(twin_path_t *path, twin_fixed_t x, twin_fixed_t y)
                      _twin_matrix_y(&path->state.matrix, x, y));
 }
 
+void twin_path_draw_polar(twin_path_t *path, twin_angle_t deg)
```
Comment (jserv) [inline src/path.c:137]: Make this function internal. i.e., declare it as a static function.

File: src/trig.c [hunk 3]
```diff
@@ -150,7 +23,18 @@ twin_fixed_t twin_sin(twin_angle_t a)
         /* mirror second and third quadrant values across y axis */
         if (a & TWIN_ANGLE_90)
             a = TWIN_ANGLE_180 - a;
-        sin = _twin_sin_table[a & (TWIN_ANGLE_90 - 1)];
+        /* 5th order polynomial approximation */
```
Comment (jserv) [inline src/trig.c:26]: Improve the comment, making it informative.

File: src/trig.c [hunk 4]
```diff
@@ -178,3 +62,10 @@ twin_fixed_t twin_tan(twin_angle_t a)
         return 0;
     return ((s << 15) / c) << 1;
 }
+
+void twin_sincos(twin_angle_t a, twin_fixed_t *sin, twin_fixed_t *cos)
+{
+    *sin = twin_sin(a);
+    *cos = twin_cos(a);
+}
+
```
Comment (jserv) [inline src/trig.c:71]: Don't leave the blank aline.

File: src/trig.c [hunk 5]
```diff
@@ -178,3 +62,10 @@ twin_fixed_t twin_tan(twin_angle_t a)
         return 0;
     return ((s << 15) / c) << 1;
 }
+
+void twin_sincos(twin_angle_t a, twin_fixed_t *sin, twin_fixed_t *cos)
```
Comment (jserv) [inline src/trig.c:66]: Here is the proofread version of your text:
Don't do that. You don't have to invoke both `twin_sin` and `twin_cos` in the function body of `twin_sincos`. Instead, you should calculate 5th order polynomial approximation in `twin_sincos`, and then both `twin_sin` and `twin_cos` should just be wrappers for `twin_sincos`.

File: src/trig.c (hunk 3 above)
Comment (jouae) [reply src/trig.c:26]: @ndsl7109256
You can refer to the comments in [vwifi](https://github.com/sysprog21/vwifi/blob/main/vwifi.c), at least providing the form of the approximate polynomial and the description of the parameters.

```c
static inline s32 __sin_s3(s32 x)
{
...
    /* S(x) = (x * (3 * 2^p - (x * x)/2^r)) / 2^s
     * @n: the angle scale
     * @A: the amplitude
     * @p: keep the multiplication from overflowing
     */
...
}
```

**Review by jouae — COMMENTED** [review]
@ndsl7109256  
Please include the number of observations on which the RMS error is based.

**Review by jserv — CHANGES_REQUESTED** [review]
Rework `twin_tan` with `twin_sincos` or approximation.

File: src/trig.c [hunk 6]
```diff
@@ -178,3 +39,64 @@ twin_fixed_t twin_tan(twin_angle_t a)
...
+}
+
+void twin_sincos(twin_angle_t a, twin_fixed_t *sin, twin_fixed_t *cos)
+{
+    twin_fixed_t sin_val = 0, cos_val = 0;
+
+    /* limit to [0..360) */
+    a = a & (TWIN_ANGLE_360 - 1);
+    int c = a > TWIN_ANGLE_90 && a < TWIN_ANGLE_270;
+    /* special case for 90 degrees - no room in table */
```
Comment (jserv) [inline src/trig.c:76]: Refine this comment since the code does not look up any tables.

**Review by jserv — CHANGES_REQUESTED** [review]
Squash the commits and refine the git commit messages.

File: src/trig.c [hunk 7]
```diff
@@ -6,167 +6,28 @@
...
 
 twin_fixed_t twin_cos(twin_angle_t a)
 {
-    return twin_sin(a + TWIN_ANGLE_90);
+    twin_fixed_t cos_val = 0;
+    twin_sincos(a, NULL, &cos_val);
+    return cos_val;
 }
 
 twin_fixed_t twin_tan(twin_angle_t a)
```
Comment (jserv) [inline src/trig.c:27]: Is it possible to calculate tan(x) without divisions?

Comment (ndsl7109256) [reply src/trig.c:27]: I think we could use Taylor series expansion to calculate tan(x) without using division, but as x approaches 90 degrees, the error would become very large. I'm not sure if this is suitable for use in this project.

Comment (jserv) [reply src/trig.c:27]: > I think we could use Taylor series expansion to calculate tan(x) without using division, but as x approaches 90 degrees, the error would become very large. I'm not sure if this is suitable for use in this project.

`TWIN_FIXED_MAX` can be set for $tan(\pi/2)$ while `TWIN_FIXED_MIN` can be set for $tan(-\pi/2)$.

Comment (ndsl7109256) [reply src/trig.c:27]: image

Though we could set $tan( \pi /2)$ to `TWIN_FIXED_MAX` directly. However, we still suffer significant error when angle is closed to $\pi/2$. Should we just neglect the error here?

Comment (jserv) [reply src/trig.c:27]: @jouae, Can you comment this? Can CORDIC algorithm be applied to tan(x)?

Comment (jserv) [reply src/trig.c:27]: > I need to spend some time studying the principles of CORDIC.

See https://github.com/Max-Gulda/Cordic-Math/blob/main/lib/cordicMath/src/cordic-math.c#L290-L332

Comment (jouae) [reply src/trig.c:27]: Perhaps we can improve the division about $\frac{x}{y}$

see https://hackmd.io/@sysprog/linux-intdiv

Comment (ndsl7109256) [reply src/trig.c:27]: > Perhaps we can improve the division about x y
> 
> see https://hackmd.io/@sysprog/linux-intdiv

@jouae Thanks for the follow-up! 
The divisor we use (x calculated in CORDIC or cos calculated in current way) varies for different angles. Is it possible to improve by the mentioned techniques? 
Could you provide a more detailed explanation of your idea? Thanks!

Comment (jouae) [reply src/trig.c:27]: @ndsl7109256 
Sure, leave it to me, dude.

**Review by jouae — APPROVED** [review]

**Review by jserv — CHANGES_REQUESTED** [review]
1. Squash the commits and refine the git commit messages.
2. Don't use markdown syntax in git commit messages. Instead, always use plain English.
3. Append "Close #13" at the end of the git commit message.
4. Add comment about CORDIC based tangent function improvement.

File: src/trig.c (hunk 7 above)
Comment (jserv) [reply src/trig.c:27]: @jouae, Create a new GitHub issue when this pull request is merged.

Comment (jouae) [reply src/trig.c:27]: @jserv 
Roger that.

**Comment by ndsl7109256** [comment]
Enhancement: Implement CORDIC algorithm for tan calculation

Currently, the `twin_tan` function uses division, which may be computationally expensive. A potential improvement would be to implement the CORDIC  algorithm for calculating tan. This would avoid the use of the divider and potentially improve performance.

References:
https://github.com/Max-Gulda/Cordic-Math/blob/main/lib/cordicMath/src/cordic-math.c#L290-L332
https://hackmd.io/@sysprog/linux-intdiv

**Comment by jserv** [comment]
Thank @ndsl7109256 for contributing!

**Pull request merged**